### PR TITLE
Mailbox locking refcount

### PR DESCRIPTION
**Pull request by brong**
This brings back the mboxlist_changesub modseq bump but also rewrites mailbox refcounting so it works, and ALSO replaces the JMAP mailbox cache as part of that change.

There are probably more places we can just open and close rather than having to track the index mailbox too.

### CODE REVIEW

File: imap/jmap_api.c
Comment (rsto) [inline imap/jmap_api.c:1061]: that comment is outdated

File: imap/jmap_api.c
Comment (rsto) [inline imap/jmap_api.c:1062]: this conditional and the `rw` variable is obsolete, this could just become

     r = req->force_openmbox_rw ? mailbox_open_iwl(name, mboxp) : mailbox_open_irl(name, mboxp);

File: imap/mboxlist.c
Comment (rsto) [inline imap/mboxlist.c:5135]: I know it's ungrateful work, but this function signature could really serve getting its bool arguments refactored in a single `flags` argument.

**Review by rsto — COMMENTED** [review]
Just have a few nits that might be nice to address, but no hard requirement. I'll wait for CI to finish to approve.

File: imap/mboxlist.c
Comment (brong) [reply imap/mboxlist.c:5135]: yes, that seems reasonable  - I will do that

File: imap/jmap_api.c
Comment (brong) [reply imap/jmap_api.c:1062]: OK?  do we ALWAYS set force_openmbox_rw?  It looks like `rw` gets set to both possible values a lot

```
8109:    int r = jmap_openmbox_by_guidrec(rock->req, rec, &mbox, /*rw*/0);
8151:            r = jmap_openmbox(req, mboxname, &mbox, 0);
8838:    r = jmap_openmbox(req, mboxname, &mbox, 1);
8982:        r = jmap_openmbox(req, dstname, &dst, 1);
12402:            if (!r) r = jmap_openmbox(req, mboxrec->mboxname, &mbox, /*rw*/1);
12477:                if (!r) jmap_openmbox(req, mbentry->name, &mbox, /*rw*/1);
13087:        int r = jmap_openmbox(req, mboxrec->mboxname, &mbox, 1);
13611:    int r = jmap_openmbox_by_guidrec(rock->req, rec, &mbox, /*rw*/1);
13742:        r = jmap_openmbox(req, mbentry->name, &mbox, 0);
13791:    r = jmap_openmbox(req, mbentry->name, &mbox, 0);
13966:            r = jmap_openmbox(req, dst_mboxname, &dst_mbox, /*rw*/1);
```

Comment (rsto) [reply imap/jmap_api.c:1062]: oops, my bad. sorry, I was wrong. so this would get ` = (rw || req->force_openmbox_rw) ? ..` then?

File: imap/mboxlist.c
Comment (brong) [reply imap/mboxlist.c:5135]: Hrm, that would change callers like this:

old:
```
mboxlist_changesub(oldname, userid, auth_state, 0, 0, 0, silent);
mboxlist_changesub(newname, userid, auth_state, 1, 0, 0, silent);
mboxlist_changesub(mbentry->name, req->userid, httpd_authstate, 0, 1, 0, 1);
```

new:
```
mboxlist_changesub(oldname, userid, auth_state, silent ? CHANGESUB_SILENT : 0);
mboxlist_changesub(newname, userid, auth_state, CHANGESUB_ADD | (silent ? CHANGESUB_SILENT : 0));
mboxlist_changesub(newname, userid, auth_state, CHANGESUB_FORCE | CHANGESUB_SILENT);

I'm not sure that's better tbh.

```

File: imap/jmap_api.c
Comment (rsto) [reply imap/jmap_api.c:1062]: Or lets get rid of `force_openmbox_rw` altogether which would also allow to remove `jmap_openmbox`

File: imap/mboxlist.c
Comment (rsto) [reply imap/mboxlist.c:5135]: yeah, let's just leave as-is then

Comment (brong) [reply imap/mboxlist.c:5135]: I left this unchanged, I think it's not worth making that change at this point

**Comment by brong** [comment]
Damn tests pass on my box :(

**Comment by rsto** [comment]
These tests are failing on my local build

     [FAILED] Cyrus::ImapTest.notify
     [FAILED] Cyrus::ImapTest.urlauth2
     [FAILED] Cyrus::ImapTest.urlauth-binary
     [FAILED] Cyrus::Info.conf
     [FAILED] Cyrus::JMAPCalendars.calendarsharenotification_changes
     [FAILED] Cyrus::JMAPCalendars.calendarsharenotification_query
     [FAILED] Cyrus::Rename.rename_inbox
     [FAILED] Cyrus::Rename.rename_inbox_unselected
     [FAILED] Cyrus::Rename.rename_bigconversation
     [FAILED] Cyrus::Rename.rename_user_bigconversation

I only glimpsed at JMAPCalendars.calendarsharenotification_changes and it seems like where creating a share notification and deleting one each bumped the modseq on the notifications mailbox, the create now silently does not bump the modseq. Haven't dug deeper.

**Comment by elliefm** [comment]
@brong If the tests pass for you locally, they might be dependent on some local config you have but no-one else does.  What's in the `[config]` section of your cassandane.ini?    The only thing that _must_ be in there these days is `zoneinfo_dir`.  If you have `altnamespace` or `unixhierarchysep` entries left over from the old days, remove them.  If you have other stuff in there that the tests are relying on, they should probably be configured in the test itself, not in your cassandane.ini

**Comment by elliefm** [comment]
This feels like a pretty high-risk PR, both due to the extent of the changes and the mysteries exposed in testing so far.

Even once this passes CI, it seems likely that other edge cases will arise that we don't yet have adequate test coverage for.  So I think this should be "DO NOT MERGE" and "include in fastmail" for a while before it lands on master, to give us time to use it for real and shake them out before we start calling it release quality.  Especially with it being only about 6-7 weeks until we fork the new 3.10 stable branch!

So I'm tagging this DO NOT MERGE for now, and if it's ready for our builds this year we can tag it "include in fastmail".  It can land on master next year, after the 3.10 branch has forked.

**Comment by brong** [comment]
Thanks @elliefm, that seems reasonable!  I'm hoping this time it will actually pass.  It's shaken out some bugs in assumptions in other code :)  But yeah, complex changes for sure.

File: imap/reconstruct.c
Comment (elliefm) [inline imap/reconstruct.c:544]: Is this leaking `mailbox`?  I don't see how it gets committed and closed after `mailbox_setversion()` now

**Review by elliefm — COMMENTED** [review]
One nit; also something odd going on in reconstruct, looks like the mailbox is neither committed nor closed after mailbox_setversion() succeeds?

File: imap/mailbox.c
Comment (brong) [reply imap/mailbox.c]: done

File: imap/reconstruct.c
Comment (brong) [reply imap/reconstruct.c:545]: fixed

**Comment by brong** [comment]
@ellie fixed the setversion issue and also added a test which exercises it!

**Comment by brong** [comment]
@rsto did you want to update your review on this?

**Comment by elliefm** [comment]
We gotta be more careful about pinging "elliefm" and "ksmurchison", rather than "ellie" and "Murch" who are nothing to do with this project ;)

This is still failing in CI, on a few JMAPContacts tests.  The failures are interesting... one was like this:

```
1) JMAPContacts.contact_blobid
 expected 'BEGIN:VCARD', got '<!DOCTYPE H' at Cassandane/Cyrus/JMAPContacts.pm line 3498.
```

Another failure expected to receive one content-type but got a completely different one (sorry, I forgot which by the time I got to typing here)

And the other got back a 500 error response, when it was expecting a document that matched a pattern like "this should not crash the server"

It kinda smells like these tests are getting back completely different documents from what they expected.  I wonder if there's something up with the changes to the jmap mailbox cache handling?

I've told the CI job to re-run, on the off-chance that it was some transient weirdness, though that seems unlikely... and tbh if it were a transient issue I think I'd be even more concerned.  I'll pull the branch and run it locally too, once my current cass run finishes, and see if that sheds any new light.

File: cassandane/Cassandane/Cyrus/Reconstruct.pm
Comment (elliefm) [inline cassandane/Cassandane/Cyrus/Reconstruct.pm:618]: what's the purpose of the `eval` here?

File: imap/mailbox.c
Comment (elliefm) [inline imap/mailbox.c:2651]: Need a comma at the end of this line, i.e. between the fixed description and the format string for the varargs -- they should be separate args, not one concatenated string.  Right now the `%s`s are part of the description, and then it's treating `mailbox_name(mailbox)` as the format string and `error_message(r)` as the one argument to it.  And since `mailbox_name(mailbox)` is a string, and it doesn't contain any `%`s whose argument types might mismatch, the compiler fails to detect the misuse. :(

**Comment by elliefm** [comment]
Here's the local failures:

```
1) JMAPContacts.contact_blobid
 expected 'BEGIN:VCARD', got '<!DOCTYPE H' at Cassandane/Cyrus/JMAPContacts.pm line 3498.
```

```
2) JMAPContacts.contact_apple_label_handling
 <!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html>
  <head>
    <title>500 Internal Server Error</title>
  </head>
  <body>
    <h1>Internal Server Error</h1>
    <p>Mailbox is locked</p>
    <hr>
    <address>Cyrus-HTTP/3.9.0-alpha0-1141-gcd87f6735 Cyrus-SASL/2.1.27 Lib/XML2.9.10 Jansson/2.13.1 OpenSSL/1.1.1n Nghttp2/1.43.0 Wslay/1.1.1 Zlib/1.2.11 Brotli/1.0.9 Zstd/1.4.8 Xapian/1.5.0 LibiCal/3.1 ICU4C/72.1 SQLite/3.34.1 Server at 127.0.0.1 Port 9112</address>
  </body>
</html>
 didn't match /(?^:X-ABLabel:this-should-not-crash-cyrus)/ at Cassandane/Cyrus/JMAPContacts.pm line 3740.
```

```
3) JMAPContacts.contact_copy
 <undef> unexpected at Cassandane/Cyrus/JMAPContacts.pm line 3292.
```

```
4) JMAPContacts.contact_set_avatar_from_deleted_contact
 expected 'image/jpeg', got 'text/html; charset=utf-8' at Cassandane/Cyrus/JMAPContacts.pm line 2471.
```

The restarted CI job also failed again, with the same four failures: https://github.com/cyrusimap/cyrus-imapd/actions/runs/6902252975/job/18830872467?pr=4731

File: imap/http_dav_sharing.c
Comment (elliefm) [inline imap/http_dav_sharing.c:387]: This xsyslog call is also missing the comma between description string and format string

```suggestion
        if (r) xsyslog(LOG_ERR, "IOERROR: failed to create notify collection",
		      "mailbox=<%s> error=<%s>",
```

**Review by elliefm — COMMENTED** [review]
Still one bad xsyslog call.  But it passes CI now! :)

**Review by elliefm — CHANGES_REQUESTED** [review]
The loop/global changes from the latest two commits look good.  But you haven't yet fixed the xsyslog typo (in http_dav_sharing) from my last review.  It'll produce a gibberish log message if the error condition it's logging about ever occurs.

**Comment by brong** [comment]
There's a ton more changes in here now!  I added namelock ordering checks and found many things.  There's still issues with:

```
Search.esearch
JMAPCalendars.calendarevent_set_itip_preserve_partstat
and a couple more that also call the deliver_local codepath
```

File: imap/mailbox.c
Comment (elliefm) [reply imap/mailbox.c:2651]: This still needs fixing

```suggestion
            xsyslog(LOG_ERR, "IOERROR: Error committing to conversations database",
```

**Review by elliefm — DISMISSED** [review]

**Review by rsto — APPROVED** [review]
LGTM

**Pull request merged**
